recorder/src/main/cpp/controller.cc

```diff
@@ -485,37 +485,39 @@ void Controller::issue_work(const std::string& host, const std::uint32_t port, s
                                 retire_work(work_id);
                             });
                     };
-                    if (w.work_size() > 0) {
+
+                    if (w.work_size() > 0) {//something actually needs to be done
                         std::uint32_t tx_timeout = cfg.slow_tx_tolerance * w.duration();
                         std::shared_ptr<HttpRawProfileWriter> raw_writer(new HttpRawProfileWriter(jvm, jvmti, host, port, raw_writer_ring, cancel_work, tx_timeout));
                         writer.reset(new ProfileWriter(raw_writer, buff));
                         recording::RecordingHeader rh;
                         populate_recording_header(rh, w, controller_id, controller_version);
                         writer->write_header(rh);
-                    }
-                    for (auto i = 0; i < w.work_size(); i++) {
-                        auto work = w.work(i);
-                        prep(work);
-                    }
 
-                    serializer.reset(new ProfileSerializingWriter(jvmti, *writer.get(), SiteResolver::method_info, SiteResolver::line_no, get_ctx_reg(), sft, tts, cfg.noctx_cov_pct));
+                        for (auto i = 0; i < w.work_size(); i++) {
+                            auto work = w.work(i);
+                            prep(work);
+                        }
 
-                    JNIEnv *env = getJNIEnv(jvm);
-                    Processes processes;
-                    for (auto i = 0; i < w.work_size(); i++) {
-                        auto work = w.work(i);
-                        issue(work, processes, env);
-                    }
+                        serializer.reset(new ProfileSerializingWriter(jvmti, *writer.get(), SiteResolver::method_info, SiteResolver::line_no, get_ctx_reg(), sft, tts, cfg.noctx_cov_pct));
```

Comment:
The transitions are right in this case aren't they? /profile call isn't necessary (because there is nothing to do).

So if for any reason backend does assign empty work, the transition should  be scheduled and finished. /profile should not be called, because there are nothing to report.

This seems like the right behavior to me.